README.md
# Example Library: RGB

**Ultra-simple library for [RGB color][rgb-wiki] representation in [Haskell][haskell], inspired by a special [Codewars][codewars] kata about a [simple greyscale conversion][kata].**

[![Build and test][test-badge-image]][test-badge-url]
[![Publish API docs][apigen-badge-image]][apigen-badge-url]
[![Haddock documentation][haddock-badge-image]][haddock-badge-url]
[![Codewars profile][codewars-badge-image]][codewars-badge-url]

## Details

This tiny library features

- Flexible types for colors and *"images"*.
- The color type has an [Applicative Functor instance][applicative] ([Read more][applicative-wiki]) that allows for component-wise modifications via *RGB functions*
- A continuous test suite with unit and property tests, covering the basics and codewars tests for greyscale images.
- Automatic [API docs][api-docs] generation via GitHub actions, hosted on GitHub pages.

## Contributors

[![Contributor Covenant 2.0][coco-badge-image]][coco-badge-url]

- Mirko Westermeier ([@memowe][memowe-gh])

## Author and License

Copyright (c) 2022 Mirko Westermeier

Released under the MIT license. See [LICENSE](LICENSE) for details.

[rgb-wiki]: https://en.wikipedia.org/wiki/RGB_color_model
[haskell]: https://www.haskell.org/
[codewars]: https://www.codewars.com/
[kata]: https://www.codewars.com/kata/590ee3c979ae8923bf00095b
[applicative]: https://hackage.haskell.org/package/base/docs/Control-Applicative.html#t:Applicative
[applicative-wiki]: https://en.wikipedia.org/wiki/Applicative_functor
[api-docs]: https://mirko.westermeier.de/Example-RGB/RGB.html
[memowe-gh]: https://github.com/memowe

[test-badge-image]: https://github.com/memowe/Example-RGB/actions/workflows/test.yml/badge.svg
[test-badge-url]: https://github.com/memowe/Example-RGB/actions/workflows/test.yml
[apigen-badge-image]: https://github.com/memowe/Example-RGB/actions/workflows/haddock-pages.yml/badge.svg
[apigen-badge-url]: https://github.com/memowe/Example-RGB/actions/workflows/haddock-pages.yml
[haddock-badge-image]: https://img.shields.io/badge/Haddock-Documentation-8a80a8?style=flat&logo=haskell&logoColor=lightgray
[haddock-badge-url]: https://mirko.westermeier.de/Example-RGB/RGB.html
[codewars-badge-image]: https://www.codewars.com/users/memowe/badges/micro?theme=light
[codewars-badge-url]: https://www.codewars.com/users/memowe
[coco-badge-image]: https://img.shields.io/badge/Code%20of%20Conduct-Contributor%20Covenant%202.0-8f761b.svg?style=flat&logo=adguard&logoColor=lightgray
[coco-badge-url]: CODE_OF_CONDUCT.md


CHANGELOG.md
# Revision history for RGB

## 0.1.0.0 -- 2022-01-21

* First version. Released on an unsuspecting world.


test/tests.hs
import Test.Tasty ( defaultMain, testGroup )

import RGB.TestRGB ( testBasics )
import RGB.TestCodewars ( testCodewars )

tests = testGroup "RGB tests"
  [ testBasics
  , testCodewars
  ]

main = defaultMain $ testGroup "Tests" [tests]


test/RGB/Test.hs
{-# LANGUAGE FlexibleInstances #-}
module RGB.Test where

import Test.QuickCheck ( choose, Arbitrary(..) )
import RGB

instance Arbitrary (RGB Int) where
  arbitrary = RGB <$> i256 <*> i256 <*> i256
    where i256 = choose (0, 255)


test/RGB/TestRGB.hs
module RGB.TestRGB (testBasics) where

import Test.Tasty ( testGroup )
import Test.Tasty.HUnit ( testCase, (@?=) )
import Test.Tasty.QuickCheck ( testProperty )
import RGB.Test
import RGB

color = RGB 17 37 42 :: RGB Int

testApplicativeFunctor = testGroup "Applicative functor"
  [ testCase "Functor" $ fmap succ color @?= RGB 18 38 43
  , testCase "Applicative" $ RGB (+1) (+2) (*2) <*> color @?= RGB 18 39 84
  ]

testListConversion = testGroup "List conversion"
  [ testProperty "RGB to list" $ \col@(RGB r g b) ->
      [r, g, b] == (toList col :: [Int])
  , testProperty "List to RGB: three elements" $ \col@(RGB r g b) ->
      fromList [r, g, b] == Just (col :: RGB Int)
  , testCase "List to RGB: two elements" $ fromList [17, 42] @?= Nothing
  ]

testStringConversion = testGroup "String conversion"
  [ testCase "Show instance" $ show color @?= "RGB {red = 17, green = 37, blue = 42}"
  , testCase "Show hex" $ showHex color @?= "#11252a"
  , testCase "Read hex: crap" $ parseRGB "crap" @?= Nothing
  , testCase "Read hex: correct" $ parseRGB "#11252a" @?= Just color
  , testProperty "parse . hex roundtrip" $ \col ->
      parseRGB (showHex col) == Just (col :: RGB Int)
  ]

testImages = testGroup "Image handling"
  [ testProperty "Image to list" $ \img ->
      toImageList img == map (map toList) img
  , testProperty "List to image" $ \img ->
      fromImageList (map (map toList) img) == Just img
  ]

testBasics = testGroup "Basic tests"
  [ testApplicativeFunctor
  , testListConversion
  , testStringConversion
  , testImages
  ]


src/RGB.hs
{-|
Module      : RGB
Description : RGB color handling library
Copyright   : (c) 2022 Mirko Westermeier
License     : MIT

RGB color handling library
-}

module RGB
  (
  -- * The RGB color type
    RGB(RGB, red, green, blue)
  -- * Utilty functions
  , toList
  , fromList
  , showHex
  , parseRGB
  -- * /Images/
  , Image
  , ImageList
  , toImageList
  , fromImageList
  -- * Codewars greyscale conversion
  , greyScale
  -- ** Greyscale images
  , color2grey
  , color2grey'
  ) where

import Data.Char ( digitToInt )
import Data.Maybe ( fromJust, listToMaybe )
import Text.Printf( printf, PrintfArg )
import Text.ParserCombinators.ReadP

-- | A triple type representing the three color components of RGB color values.
-- Since most systems expect component values from 0 to 255, unsigned integer
-- types like @'RGB' 'Word'@ or @'RGB' 'Int'@ are most useful for color values.
-- But keeping it parameterized here allows for a useful 'Applicative' instance
-- (see below).
data RGB a = RGB  { red   :: a -- ^ Red color value
                  , green :: a -- ^ Green color value
                  , blue  :: a -- ^ Blue color value
                  }
  deriving
    ( Eq
    -- ^ Default 'Eq' instance. Two 'RGB' values are equal /iff/ all components
    -- 'red', 'green', 'blue' are equal.
    , Show
    -- ^ Default 'Show' instance. Example: @"RGB {red = 17, green = 37, blue = 42}"@.
    -- See 'showHex' for a compressed hexadecimal stringification.
    )

-- | Application of a given function to all color components:
--
-- prop> fmap (*2) (RGB 2 7 21)  ==  RGB 4 14 42
instance Functor RGB where
  fmap = (<*>) . pure

-- | The 'Applicative' instance of 'RGB' can act as a component-wise application:
--
-- prop> (RGB (+5) (`div` 2) (*6)  <*>  RGB 12 74 7)  ==  RGB 17 37 42
instance Applicative RGB where
  pure v = RGB v v v
  (RGB fr fg fb) <*> (RGB r g b) = RGB (fr r) (fg g) (fb b)

-- | Creates a three element list of 'RGB' color components:
--
-- prop> toList (RGB 17 37 42)  ==  [17, 37, 42]
toList :: RGB a -> [a]
toList (RGB r g b) = [r, g, b]

-- | Tries to create a color from a list of values, assuming exactly three values:
--
-- prop> fromJust (fromList [17, 37, 42])  ==  RGB 17 37 42
fromList :: [a] -> Maybe (RGB a)
fromList [r, g, b]  = Just $ RGB r g b
fromList _          = Nothing

-- | Hexadecimal color representation for 'printf'able components. For
-- values between 0 and 255 it is the inverse of 'parseRGB' (modulo 'Maybe'):
--
-- prop> showHex (fromJust $ parseRGB $ "#11252a")  ==  "#11252a"
showHex :: PrintfArg a => RGB a -> String
showHex (RGB r g b) = printf "#%02x%02x%02x" r g b

-- | Parse @'RGB' 'Int'@ values from hexadecimal strings like @#aabbcc@.
--
-- prop> fromJust (parseRGB $ showHex $ RGB 17 37 42)  ==  RGB 17 37 42
parseRGB :: String -> Maybe (RGB Int)
parseRGB = fmap fst . listToMaybe . reverse . readP_to_S parse
  where parse = do  char '#'
                    r <- comp
                    g <- comp
                    b <- comp
                    return $ RGB r g b
        comp  = do  d1 <- digit
                    d2 <- digit
                    return $ d1 * 16 + d2
        digit = do  digChar <- satisfy (`elem` "0123456789abcdef")
                    return $ digitToInt digChar

-- | Super-simple image representation as a two-dimensional list of /pixels/
-- of type @'RGB' a@.
type Image = [[RGB Int]]

-- | Super-oversimplified image representation as a three-dimensional list
-- of color components of type 'Int'.
type ImageList = [[[Int]]]

-- | Creates an 'ImageList' from an 'Image'.
toImageList :: Image -> ImageList
toImageList = map $ map toList

-- | Tries to create an 'Image' from an 'ImageList' using 'fromList'.
fromImageList :: ImageList -> Maybe Image
fromImageList = mapM $ mapM fromList

-- | Simple greyscale conversion of a @'RGB' 'Int'@ color according to the
-- [Codewars](https://codewars.com) kata
-- __[Convert Color image to greyscale](https://www.codewars.com/kata/590ee3c979ae8923bf00095b)__:
--
-- @
-- greyScale (RGB r g b)  ==  RGB ((r+g+b)\/3) ((r+g+b)\/3) ((r+g+b)\/3)
-- @
-- but using 'Fractional' arithmetics.
greyScale :: RGB Int -> RGB Int
greyScale (RGB r g b) = let compSum = fromIntegral (r+g+b)
                            average = round (compSum / 3)
                        in  pure average

-- | Application of 'greyScale' to each @'RGB' 'Int'@ pixel of an 'Image'
color2grey :: Image -> Image
color2grey = map $ map greyScale

-- | Application of 'greyScale' to each @'RGB' 'Int'@ pixel of an 'ImageList'
color2grey' :: ImageList -> Maybe ImageList
color2grey' = fmap (toImageList . color2grey) . fromImageList
